Marvelys-Torres/tarea-5/index.js

```diff
@@ -0,0 +1,201 @@
+class Caja {
+  constructor(x, y) {
+    this.x = x;
+    this.y = y;
+  }
+
+  getCajaArriba() {
+    if (this.y === 0) return null;
+    return new Caja(this.x, this.y - 1);
+  }
+
+  getCajaDerecha() {
+    if (this.x === 3) return null;
+    return new Caja(this.x + 1, this.y);
+  }
+
+  getCajaAbajo() {
+    if (this.y === 3) return null;
+    return new Caja(this.x, this.y + 1);
+  }
+
+  getCajaIzquierda() {
+    if (this.x === 0) return null;
+    return new Caja(this.x - 1, this.y);
+  }
+
+  getCajasVecinas() {
+    return [
+      this.getCajaArriba(),
+      this.getCajaDerecha(),
+      this.getCajaAbajo(),
+      this.getCajaIzquierda(),
+    ].filter((caja) => caja !== null);
+  }
+
+  getCajaVecinaAleatoria() {
+    const cajasVecinas = this.getCajasVecinas();
+    return cajasVecinas[Math.floor(Math.random() * cajasVecinas.length)];
+  }
+}
+
+const intercambiarCajas = (rejilla, caja1, caja2) => {
+  const temporal = rejilla[caja1.y][caja1.x];
+  rejilla[caja1.y][caja1.x] = rejilla[caja2.y][caja2.x];
+  rejilla[caja2.y][caja2.x] = temporal;
+};
+
+const estaResuelto = (rejilla) => {
+  return (
+    rejilla[0][0] === 1 &&
+    rejilla[0][1] === 2 &&
+    rejilla[0][2] === 3 &&
+    rejilla[0][3] === 4 &&
+    rejilla[1][0] === 5 &&
+    rejilla[1][1] === 6 &&
+    rejilla[1][2] === 7 &&
+    rejilla[1][3] === 8 &&
+    rejilla[2][0] === 9 &&
+    rejilla[2][1] === 10 &&
+    rejilla[2][2] === 11 &&
+    rejilla[2][3] === 12 &&
+    rejilla[3][0] === 13 &&
+    rejilla[3][1] === 14 &&
+    rejilla[3][2] === 15 &&
+    rejilla[3][3] === 0
+  );
+};
+
+const getRejillaAleatoria = () => {
+  let rejilla = [
+    [1, 2, 3, 4],
+    [5, 6, 7, 8],
+    [9, 10, 11, 12],
+    [13, 14, 15, 0],
+  ];
+  let cajaBlanca = new Caja(3, 3);
+  for (let i = 0; i < 1000; i++) {
+    const cajaVecinaAleatoria = cajaBlanca.getCajaVecinaAleatoria();
+    intercambiarCajas(rejilla, cajaBlanca, cajaVecinaAleatoria);
+    cajaBlanca = cajaVecinaAleatoria;
+  }
+  if (estaResuelto(rejilla)) return getRejillaAleatoria();
+  return rejilla;
+};
+
+class Estado {
+  constructor(rejilla, movimiento, tiempo, estado) {
+    this.rejilla = rejilla;
+    this.movimiento = movimiento;
+    this.tiempo = tiempo;
+    this.estado = estado;
+  }
+
+  static listo() {
+    return new Estado(
+      [
+        [0, 0, 0, 0],
+        [0, 0, 0, 0],
+        [0, 0, 0, 0],
+        [0, 0, 0, 0],
+      ],
+      0,
+      0,
+      "listo"
+    );
+  }
+
+  static empezar() {
+    return new Estado(getRejillaAleatoria(), 0, 0, "jugando");
+  }
+}
+
+class Juego {
```

Comment:
@MarvelysTorres each CLass should be on its own file, like game.js also should be named in english